### ISSUE TITLE
support sqlalchemy 2.0

### ISSUE THREAD

**Issue by Brooke-white** [sqlalchemy-redshift/sqlalchemy-redshift#264]
sqlalchemy 2.0 was released earlier this month. https://docs.sqlalchemy.org/en/20/dialects/ we should work to support it. 

- read sqlalchemy's docs to understand breaking changes between 1.4 and 2.0
- identify what parts of our existing code need to be refactored to support 2.0 while maintaining support for 1.4
- identify what new 2.0 features need support in our dialects
- expand upon tests based on new apis in 2.0

**Comment by stuck1233333** [comment]
Is there an estimated time frame for supporting this release? Thank you!

**Comment by Brooke-white** [comment]
hey @stuck1233333, I do not have an ETA for when we will support but am happy to support any folks wishing to help add support (e.g. troubleshooting, running changes against CI, etc.). In parallel, I'm discussing with the Redshift team to see when we can find bandwidth for this on our roadmap.

**Comment by matthewgdv** [comment]
So, I've been working on adding an override hook [upstream](https://github.com/sqlalchemy/sqlalchemy/pull/9475) in `sqlalchemy` to fix the [regression caused in 2.0](https://github.com/sqlalchemy/sqlalchemy/issues/9442) which causes a `ProgrammingError` to be thrown in `PGDialect.initialize()` against Redshift.

Because of how I've done it upstream the fix on our side should be as simple as implementing a method in the redshift dialect that just sets backslash escapes to `False`. This will work on 1.4 because it simply won't be called but the special-cased logic to support redshift is still present upstream, and it will work against the upcoming version of sqlalchemy2 once my changes are merged in and released.

I don't know if there are any other regressions or issues at the moment since that one prevents even instanciating an engine, but at least it's a start. I'll open a PR for this later today.

**Comment by Brooke-white** [comment]
thanks so much @matthewgdv  -- I'm working through our PR on spectrum support, but will take a look at your PR + evaluate other potentially regressions for migrating to sqlalchemy 2.0 after that wraps up 😄

**Comment by ryanjoycemacq** [comment]
Hello - I saw this issue was still open and was just curious if there was any update to a timeline for 2.0 support. Thank you so much.

**Comment by helensilva14** [comment]
> Hello - I saw this issue was still open and was just curious if there was any update to a timeline for 2.0 support. Thank you so much.

I was wondering the same, friendly adding other comment about it just to have the issue updated, thanks a lot!

**Comment by NelsonAndrew** [comment]
We are also interested in a timeline for sqlalchemy 2.0 support - it is holding back our adoption.

**Comment by cilantro-ditrek** [comment]
I'll hop in here as well -- my team's watching this issue closely so we can upgrade SQLAlchemy to 2.0

**Comment by helensilva14** [comment]
An idea: maybe we can use the comments section of this post to reach out to the developers? https://aws.amazon.com/blogs/big-data/use-the-amazon-redshift-sqlalchemy-dialect-to-interact-with-amazon-redshift/

**Comment by j-adamczyk** [comment]
Any news on this?

**Comment by thomykut** [comment]
We are looking for this support too!

**Comment by wellingtonf-souza** [comment]
Is there any perspective on when this will be supported?

**Comment by SebastianSchroeder** [comment]
My team also wants to upgrade to SQLALchemy 2.0. It would be great if you could find the time to work on this.

**Comment by Brooke-white** [comment]
hi folks, we are evaluating our roadmap plans for 2024 and I am requesting this feature is prioritized. I'll provide an update once we close on a plan.

**Comment by DanCardin** [comment]
Fwiw, I've been developing against https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/293 locally for the last couple of days (largely because it was a preexisting fork without the `<2` version constraint, that seems to be fixing a bug), and haven't encountered any issues. However if I ignore the version constraint, the current release seems to work just as well.

Not to say that there aren't incompatibilities still for non-vanilla-querying, but that from my perspective, the version constraint is doing more harm than good since the last few releases which have already improved compatibility.

As not-a-maintainer, it seems like it'd be more useful to remove the constraint, release that, and then deal with incompatibility bugs as they show up, rather than this more nebulous issue that doesn't indicate what might not actually be working.

**Comment by stefanks** [comment]
@Brooke-white is there an update on this issue?

**Comment by chirayu-miles** [comment]
Our codebase is on SA 2.0 now.  Stuck unable to use psycopg2 (which throws a `ProgrammingError: (psycopg2.errors.UndefinedObject) unrecognized configuration parameter "standard_conforming_strings"`) and `sqlalchemy-redshift` (which doesn't support SA 2.0.)

Any timeline for this?

**Comment by vnktsh** [comment]
I started using this repo when AWS suggested this library in one of their articles. I have reached out to AWS team to remove all references to this library; it's been in maintenance mode for an year, no support for security-robust features of SQLAlchemy 2+ version and owners don't seem to merge PRs either.

**Comment by j-adamczyk** [comment]
@vnktsh I absolutely agree. We removed `sqlalchemy-redshift` from our code, and instead use: SQLAlchemy 2.*, psycopg 3.* and redshift-connector 2.*, and it works perfectly.

**Comment by stefanks** [comment]
Unfortunately redshift-connector doesn't work for me :(
https://github.com/aws/amazon-redshift-python-driver/issues/214

**Comment by wellingtonf-souza** [comment]
@j-adamczyk how did you define the url for connection?

**Comment by j-adamczyk** [comment]
@wellingtonf-souza the whole section is very simple:
```
    conn = redshift_connector.connect(
        host=config.DB_REDSHIFT_HOST,
        port=config.DB_REDSHIFT_PORT,
        database=config.DB_REDSHIFT_DATABASE,
        user=config.DB_REDSHIFT_USERNAME,
        password=config.DB_REDSHIFT_PASSWORD,
    )

    cursor = conn.cursor()
    cursor.execute(redshift_sql)
    targets = cursor.fetchall()
```
Read things from configuration, execute SQL read from file as a string, fetch all results (assuming they fit in memory).

**Comment by dlaflamme** [comment]
In the interest of getting #293 over the finish-line, I, like @DanCardin, have run a test against #293 and verified that for a simple query it works and gets rid of the `unrecognized configuration parameter "standard_conforming_strings"` error.

This is using `SQLAlchemy==2.0.30` 

There IS a warning from SA:

```
/home/dlaflamme/test.py:13: SADeprecationWarning: The dbapi() classmethod on dialect classes has been renamed to import_dbapi().  Implement an import_dbapi() classmethod directly on class <class 'sqlalchemy_redshift.dialect.RedshiftDialect_psycopg2'> to remove this warning; the old .dbapi() classmethod may be maintained for backwards compatibility.
```

but that can be worked around by changing 2 classmethod names from `dbapi` to `import_dbapi` in `sqlalchemy_redshift/dialect.py` as the warning suggests.

I also needed to add a `connect_args={'sslmode': 'verify-ca'}` as an argument to `create_engine()` else I got the following:

```
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server certificate for "*.????????????.us-east-1.redshift.amazonaws.com" does not match host name "redshift.mycompany.com" 
```

where `redshift.mycompany.com` is a CNAME to the customer-specific `redshift.amazonaws.com` hostname above. It is unclear to me whether this is something that should be set in `sqlalchemy-redhshift` or if it should be explicitly set by the user, as I did here to fix the error. For what it is worth, the sslmode connect_arg was not required in when using `postgresql+psycopg2://` URLs with sqlalchemy 1.4.x.

**Comment by wellingtonf-souza** [comment]
We were able to use SQLAlchemy 2.0.30 with Redshift creating the engine as follows:
~~~python
engine = create_engine(f'postgresql+psycopg2://{username}:{password}@{host}:5439/{dbname}')
if hasattr(engine.dialect, "_set_backslash_escapes"):
    engine.dialect._set_backslash_escapes = lambda _: None
~~~

**Comment by bclaud** [comment]
> We were able to use SQLAlchemy 2.0.30 with Redshift creating the engine as follows:
> 
> ```python
> engine = create_engine(f'postgresql+psycopg2://{username}:{password}@{host}:5439/{dbname}')
> if hasattr(engine.dialect, "_set_backslash_escapes"):
>     engine.dialect._set_backslash_escapes = lambda _: None
> ```

Thank you, that workout really helped me solve my issue. I just needed a simple select.

**Comment by nlm4145** [comment]
> We were able to use SQLAlchemy 2.0.30 with Redshift creating the engine as follows:
> 
> ```python
> engine = create_engine(f'postgresql+psycopg2://{username}:{password}@{host}:5439/{dbname}')
> if hasattr(engine.dialect, "_set_backslash_escapes"):
>     engine.dialect._set_backslash_escapes = lambda _: None
> ```

This isn't an all encompassing solution. for example, you can't use the MetaData object from sqlalchemy. Just pointing this out to not detract from the fact that this library severely needs updating.

```
from sqlalchemy import create_engine, MetaData 
engine = create_engine(f'postgresql+psycopg2://{username}:{password}@{host}:5439/{dbname}')
if hasattr(engine.dialect, "_set_backslash_escapes"):
    engine.dialect._set_backslash_escapes = lambda _: None

meta = MetaData()
meta.reflect(bind=engine, schema=db)

sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedColumn) column pg_class.relpersistence does not exist
[SQL: SELECT pg_catalog.pg_class.relname 
FROM pg_catalog.pg_class JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace.oid = pg_catalog.pg_class.relnamespace 
WHERE pg_catalog.pg_class.relkind = ANY (ARRAY[%(param_1)s, %(param_2)s]) AND pg_catalog.pg_class.relpersistence != %(relpersistence_1)s AND pg_catalog.pg_namespace.nspname = %(nspname_1)s]
[parameters: {'param_1': 'r', 'param_2': 'p', 'relpersistence_1': 't', 'nspname_1': 'assets'}]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

**Comment by Jonasrg** [comment]
Just want to add that pandas [updated their required SQLAlchemy](https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#increased-minimum-versions-for-dependencies) version to 2.0.0 in January with their 2.2.0 release. Since the packages are likely often used in conjunction, support for SQLAlchemy 2.0 would be much appreciated.

**Comment by nlm4145** [comment]
@Brooke-white can we please get a timeline on this?

**Comment by joarobles** [comment]
is this even being considered at the time?

**Comment by DuncBegg** [comment]
for what it is worth, i have posted a question about this on the AWS re:post.

https://repost.aws/questions/QUJaXpvGmdQq2_YfXffMzJaQ/amazon-redshift-dialect-for-sqlalchemy
Upvote it if you have a second....

**Comment by nlm4145** [comment]
> for what it is worth, i have posted a question about this on the AWS re:post.
> 
> [https://repost.aws/questions/QUJaXpvGmdQq2_YfXffMzJaQ/amazon-redshift-dialect-for-sqlalchemy](url) Upvote it if you have a second....

I upvoted, commented, and flagged the support teams' response as it seems they have misinterpreted your post. I have also attempted to reach out to brooke white on LI a couple of weeks back but nothing yet.

**Comment by leodrivera** [comment]
Hi folks
I've been using sqlalchemy within [Great Expectations](https://github.com/great-expectations/great_expectations) and I could only make it work after applying this solution to the sqlalchemy code:
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4496/1/lib/sqlalchemy/dialects/postgresql/base.py

From what was described in https://github.com/sqlalchemy/sqlalchemy/issues/9442, there was a regression of https://github.com/sqlalchemy/sqlalchemy/issues/5698 in 2.0.0

Unfortunately, the implemented solution in https://github.com/sqlalchemy/sqlalchemy/pull/9475 and https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/266 did not work for me.

Update: after updating Great Expectations to 1.0,  it seemed to solve the issue for me

**Comment by oliver-helix** [comment]
Any update on support fort his issue?

**Comment by maoxingda** [comment]
+1

**Comment by SebastianSchroeder** [comment]
We reached out to our aws key account manager and he said we should not expect an update in 2024.

**Comment by nlm4145** [comment]
I have created an issue in SQLAlchemy pleading for help & support. If you could reply / react to that issue maybe it will garner more attention. Additionally, if there are other features you have noticed are broken when trying to use this dialect w/ SA 2.0 please share that as well.

[SQLAlchemy Issue](https://github.com/sqlalchemy/sqlalchemy/issues/11949)

**Comment by dsedivec** [comment]
> I have created an issue in SQLAlchemy pleading for help & support.

They closed it, and FWIW I think they're right: this should be AWS's problem.  Redshift seems to have simply terrible client support.  The JDBC driver presumably works, and _maybe_ their ODBC driver—I've never used either—but everything else is rotting from my perspective.  I can't use Redshift with all sorts of popular tools that connect to PostgreSQL, including current SQLAlchemy, Azure Data Studio, pgAdmin...  I've given up.  I have to use an old version of the `psql` CLI to get completion working.  I can't remember if psycopg2 works.

I'm definitely avoiding Redshift for future projects due in part to this general state of neglect.

**Comment by noamelf** [comment]
We are also waiting on this feature to be able to migrate to sqlalchemy 2.x

**Comment by calebAtIspot** [comment]
**Summary:**

* [A fix](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/266) has already been merged in for the `unrecognized configuration parameter "standard_conforming_strings"` error that happens with sqlalchemy v2
* PR #293 gives partial v2 compatibility, but lacks full compatibility, so it has not been merged in yet. If someone is really interested in fixing the issue they could submit a PR with full compatibility.
* You could get sqlalchemy-redshift partially working with one of the hacks in the above thread, but it's not recommended. A better idea would be to either A) downgrade sqlalchemy to a pre-v2 version like`1.4.53` B) switch away from sqlalchemy and sqlalchemy-redshift to the official Amazon [redshift_connector](https://docs.aws.amazon.com/redshift/latest/mgmt/python-redshift-driver.html) library or C) submit a PR to fix this issue (see above bullet point)

_extra notes_ 

A) sqlalchemy v1 has still been getting updates, it even supports python 3.13. However, the latest version of pandas no longer supports sqlalchemy v1. You would also need to pin the pandas version to `2.1.4` or [change your code](https://stackoverflow.com/questions/38332787/pandas-to-sql-to-sqlite-returns-engine-object-has-no-attribute-cursor). 
B) technically you can use redshift_connector _with_ this library to avoid needing to change your code, but then you still have the same problem of this library not being compatible with sqlalchemy v2. If you use redshift_connector best just to drop this library and change your code to redshift_connector's custom style.

**Comment by nlm4145** [comment]
@calebAtIspot it's mind boggling that AWS has prioritized support for redshift connector over sqlalchemy redshift dialect. like why are you re-inventing the wheel and creating tech debt? every other major data warehouse just keeps their sqlalchemy dialects updated and call it a day.

**Comment by VolkovGeoPhy** [comment]
@SebastianSchroeder they said something like this in 2024.. do not expect update in 2024..
Now I created another ticket. I noticed that they don't support NUMERIC type.. So this time (in 2025) they just suggested to checked updates for Redshift.. It makes no sense.. I could get a clear response unfortunately

**Comment by resatcaner-sp** [comment]
any updates on this matter? We have upcoming projects where we want to utilize sqlalchemy>2.0 and use redshift among other dbs that we have

**Comment by edgarrmondragon** [comment]
Has someone tried https://github.com/juiceinc/sqlalchemy-redshift?

It advertises

> Amazon Redshift SQLAlchemy Dialect with SQLAlchemy 2 support

**Comment by ArkFreestyle** [comment]
August 2025, still struggling! If the devs aren't going to update it, can we use this thread to discuss workarounds?

**Comment by torext** [comment]
> Has someone tried https://github.com/juiceinc/sqlalchemy-redshift?
> 
> It advertises
> 
> > Amazon Redshift SQLAlchemy Dialect with SQLAlchemy 2 support 
> 

I tried it and, apart from spurious warnings regarding outdated setuptools usage, reflection also still doesn't seem to work. In other words, for me it behaves no better than the simple dialect hack mentioned way above that at least makes select statements work but leaves the MetaData object unusable.

Interacting with both PostgeSQL DBs and a Redshift DBs at work I have to say this is a sorry state of affairs; I'm wholeheartedly discouraging anyone I talk to from using or even considering using Redshift for any projects whatsoever. This kind crappy support/software should not be tolerated in this day and age...

**Comment by DuncBegg** [comment]
Kind of agree. Redshift has been a genuine disappointment. The latest thing i've discovered is that it's virtually impossible to hook up redshift resultset returning stored procedures to Tableau.

**Comment by amotl** [comment]
Hi again. @awsmadi submitted a modernization patch, thank you very much. We hope it will be merged soon, followed by a package release on PyPI.

- https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/327

**Comment by qixotic** [comment]
Has anyone seen the work on this fork and have the expertise to evaluate it: https://github.com/vahid110/sqlalchemy-redshift/tree/sqlalchemy2? Perhaps @vahid110 (if you see this!) could comment on the effort.

**Comment by vahid110** [comment]
Hi @qixotic 

That’s correct. I’m currently working on the migration to SQLAlchemy 2.x.

The branch you saw (https://github.com/vahid110/sqlalchemy-redshift/tree/sqlalchemy2) was exploratory work and is nearly complete. In parallel, I’m developing a revised implementation that leverages redshift_connector more extensively under the hood. This approach relies more on SHOW commands rather than querying various svv* and pg* system tables.

Once finalised, this will become the official AWS-supported version.

**Comment by amotl** [comment]
Hi again,

the patch submitted by @TeddyCr was merged last month. Thank you so much!

- https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/326

However, we haven't been able to spot a package published to PyPI that includes the improvements.

- https://pypi.org/project/sqlalchemy-redshift/

Are you planning the next release in one way or another? It would be so sweet if we could use the Python package, and finally leave SQLAlchemy 1.x behind. 🙏 

With kind regards,
Andreas.

**Comment by TeddyCr** [comment]
closing this has solve in 1.0.0 release

**Comment by torext** [comment]
> closing this has solve in 1.0.0 release

Unfortunately reflection still doesn't seem to work properly in conjunction with `psycopg2`. Below the error when attempting to reflect a table from a Redshift DB using `sqlacodegen`:

```
ProgrammingError: (psycopg2.errors.UndefinedColumn) column pg_class.relpersistence does not exist

[SQL: SELECT pg_catalog.pg_class.relname 
FROM pg_catalog.pg_class JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace."oid" = pg_catalog.pg_class.relnamespace 
WHERE pg_catalog.pg_class.relkind = ANY (ARRAY[%(param_1)s]) AND pg_catalog.pg_class.relpersistence != %(relpersistence_1)s AND pg_catalog.pg_namespace.nspname = %(nspname_1)s]
[parameters: {'param_1': 'm', 'relpersistence_1': 't', 'nspname_1': 'platform'}]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

Note that the DB URL is correctly constructed as `redshift+psycopg2://user@host.com:port/dbname`
